src/platform/file_manager.c

```diff
@@ -0,0 +1,220 @@
+    wchar_t *result = (wchar_t *) malloc(sizeof(wchar_t) * size_needed);
+    MultiByteToWideChar(CP_UTF8, 0, str, -1, result, size_needed);
+    return result;
+}
+
+#else // not _WIN32
+#define fs_dir_type DIR
+#define fs_dir_entry struct dirent
+#define fs_dir_open opendir
+#define fs_dir_close closedir
+#define fs_dir_read readdir
+#define dir_entry_name(d) ((d)->d_name)
+#define dir_entry_close_name(n)
+#endif
+
+#ifndef S_ISLNK
+#define S_ISLNK(m) 0
+#endif
+
+#ifndef S_ISSOCK
+#define S_ISSOCK(m) 0
+#endif
+
+#ifdef __vita__
+#define CURRENT_DIR VITA_PATH_PREFIX
+#elif defined(_WIN32)
+#define CURRENT_DIR L"."
+#else
+#define CURRENT_DIR "."
+#endif
+
+#ifdef _MSC_VER
+#include <direct.h>
+#define chdir _chdir
+#elif !defined(__vita__)
+#include <unistd.h>
+#endif
+
+int platform_file_manager_get_directory_contents_by_extension(char ***file_list_ptr, int *file_count, const char *extension, int max_files)
+{
+    char **file_list = *file_list_ptr;
+    fs_dir_type *d = fs_dir_open(CURRENT_DIR);
+    if (!d) {
+        return 0;
+    }
+    int num_files = 0;
+    fs_dir_entry *entry;
+    struct stat file_info;
+    while ((entry = fs_dir_read(d))) {
+        const char *name = dir_entry_name(entry);
+        if (stat(name, &file_info) != -1) {
+            int m = file_info.st_mode;
+            if (S_ISDIR(m) || S_ISCHR(m) || S_ISBLK(m) || S_ISFIFO(m) || S_ISSOCK(m)) {
+                continue;
+            }
+        }
+        if (file_has_extension(name, extension)) {
+            if (num_files >= max_files) {
+                max_files = dir_expand_listing_size();
+                file_list = *file_list_ptr;
+            }
+            strncpy(file_list[num_files], name, FILE_NAME_MAX);
+            file_list[num_files][FILE_NAME_MAX - 1] = 0;
+            ++num_files;
+        }
+        dir_entry_close_name(name);
+    }
+    fs_dir_close(d);
+    *file_count = num_files;
+    return 1;
+}
+
+int platform_file_manager_set_base_path(const char *path)
+{
+    if (!path) {
+        log_error("set_base_path: path was not set. Julius will probably crash.", 0, 0);
+        return 0;
+    }
+    return chdir(path) == 0;
+}
+
+static int correct_case(const char *dir, char *filename)
+{
+    // Note: we do not use the _w* variants for Windows here, because the
+    // Windows filesystem is case insensitive and doesn't need corrections
+    DIR *d = opendir(dir);
+    if (!d) {
+        return 0;
+    }
+    struct dirent *entry;
+    while ((entry = readdir(d))) {
+        if (string_compare_case_insensitive(entry->d_name, filename) == 0) {
+            strcpy(filename, entry->d_name);
+            closedir(d);
+            return 1;
+        }
+    }
+    closedir(d);
+    return 0;
+}
+
+static void move_left(char *str)
+{
+    while (*str) {
+        str[0] = str[1];
+        str++;
+    }
+    *str = 0;
+}
+
+const char *platform_file_manager_case_correct_file(const char *filepath)
```

Comment:
This change is needed for Android, that's why I already added it to `file_manager`.

Reply:
Why is it necessary for Android?
Also, can we perhaps wait with this change until the language selector has been implemented? That one also heavily modifies this function, and those are logic changes that I'd like to have in the core files, not in platform-dependent ones.

Reply:
> Why is it necessary for Android?

Android doesn't use the regular filesystem functions, so the usual case correction functions don't work for it.

To compensate, all the case correction is done on the Java side when opening a file.

So for Android, this function is unneeded and should do nothing, so to avoid adding an `#ifdef __ANDROID__` to `dir.c`, I separated the function, making this part platform-dependant.

> Also, can we perhaps wait with this change until the language selector has been implemented?

Do you mean this specific change to `platform_file_manager_case_correct_file` or to the whole PR?

Reply:
> Do you mean this specific change to platform_file_manager_case_correct_file or to the whole PR?

This specific part only.

Reply:
So while restoring the old code, I remembered that now that `platform_file_manager_list_directory_contents` has a callback, `correct_case` can be made redundant. If we simply add a `dir` parameter to the `list_directory_contents` function, it can list the files which can then be checked for proper casing directly from the callback in `dir.c`.

Also, the callback can return `0` if the next file in the dir list should be checked, or `1` if the loop should be broken.

Later on I'll adapt things for Android.

What do you think?

What do you think?

Reply:
Callback returning 0 or 1 sounds good, but perhaps reverse the meaning: 0 = stop, 1 = continue with the next file?

I'm fine with the dir parameter, pass `NULL` to use current working dir?